netket/graph/custom_graph.py

```diff
@@ -0,0 +1,99 @@
+from abstract_graph import PyAbstractGraph
+
+import numpy as _np
+import networkx as _nx
+
+class PyCustomGraph(PyAbstractGraph):
+	r"""A custom graph, specified by a list of edges and optionally colors."""
+
+	def __init__(self, edges=[]):
+		"""
+		Constructs a new graph given a list of edges.
+
+		Args:
+		    edges: If `edges` has elements of type `Tuple[int, int]` it is treated
+		        as a list of edges. Then each element `(i, j)` means a connection
+		        between sites `i` and `j`. It is assumed that `0 <= i <= j`. Also,
+		        `edges` should contain no duplicates. If `edges` has elements of
+		        type `Tuple[int, int, int]` each element `(i, j, c)` represents an
+		        edge between sites `i` and `j` colored into `c`. It is again assumed
+		        that `0 <= i <= j` and that there are no duplicate elements in `edges`.
+
+
+		Examples:
+		    A 10-site one-dimensional lattice with periodic boundary conditions can be
+		    constructed specifying the edges as follows:
+
+		    >>> import netket
+		    >>> g=netket.graph.PyCustomGraph([[i, (i + 1) % 10] for i in range(10)])
+		    >>> print(g.n_sites)
+		    10
+		"""
+
+		assert isinstance(edges, list)
```

Comment:
This raises an `AssertionError`, which is different fro the `ValueError` raised by the checks below.

Reply:
Which one should I use? Should I only raise ValueErrors?

Reply:
Hm, strictly speaking it should probably a `TypeError` for this specific case, see https://docs.python.org/3/library/exceptions.html#TypeError.

Reply:
@femtobit I have implemented these errors as `TypeError`